File: orion-viewer/src/main/java/universe/constellation/orion/viewer/OrionHelpActivity.java
/*
 * Orion Viewer - pdf, djvu, xps and cbz file viewer for android devices
 *
 * Copyright (C) 2011-2013  Michael Bogdanov & Co
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package universe.constellation.orion.viewer;

import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

/**
 * User: mike
 * Date: 26.12.11
 * Time: 15:08
 */
public class OrionHelpActivity extends OrionBaseActivity {

    public static class InfoFragment extends Fragment {
        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.general_help, container, false);
        }

        @Override
        public void onActivityCreated(Bundle savedInstanceState) {
            super.onActivityCreated(savedInstanceState);
            ImageButton btn = (ImageButton) getActivity().findViewById(R.id.help_close);
            btn.setOnClickListener(new View.OnClickListener() {
                public void onClick(View v) {
                    getActivity().finish();
                }
            });
        }
    }


    public static class AboutFragment extends Fragment {
        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            View view = inflater.inflate(R.layout.about, container, false);
            TextView viewById = (TextView)view.findViewById(R.id.about_version_name);
            viewById.setText(BuildConfig.VERSION_NAME);
            return view;
        }

        @Override
        public void onActivityCreated(Bundle savedInstanceState) {
            super.onActivityCreated(savedInstanceState);
            ImageButton btn = (ImageButton) getActivity().findViewById(R.id.info_close);
            btn.setOnClickListener(new View.OnClickListener() {
                public void onClick(View v) {
                    getActivity().finish();
                }
            });
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onOrionCreate(savedInstanceState, R.layout.file_manager);
        initHelpScreen();
    }

    protected void initHelpScreen() {
        HelpSimplePagerAdapter pagerAdapter = new HelpSimplePagerAdapter(getSupportFragmentManager(), 2);
        ViewPager viewPager = (ViewPager) findViewById(R.id.viewpager);
        viewPager.setAdapter(pagerAdapter);
        TabLayout tabLayout = (TabLayout) findViewById(R.id.sliding_tabs);
        tabLayout.setupWithViewPager(viewPager);

        TabLayout.Tab help = tabLayout.getTabAt(0);
        if (help != null) {
            help.setIcon(R.drawable.help);
        }
        TabLayout.Tab about = tabLayout.getTabAt(1);
        if (about != null) {
            about.setIcon(R.drawable.info);
        }
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.file_manager_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.exit_menu_item:
                finish();
                return true;
        }
        return false;
    }

    @Override
    public boolean supportDevice() {
        return false;
    }
}


class HelpSimplePagerAdapter extends FragmentStatePagerAdapter {
    private final int pageCount;

    public HelpSimplePagerAdapter(FragmentManager fm, int pageCount) {
        super(fm);
        this.pageCount = pageCount;
    }

    @Override
    public Fragment getItem(int i) {
        return i == 0 ?
                new OrionHelpActivity.InfoFragment() : new OrionHelpActivity.AboutFragment();
    }

    @Override
    public int getCount() {
        return pageCount;
    }

}

File: orion-viewer/src/main/java/universe/constellation/orion/viewer/OrionHelpActivity.kt
/*
 * Orion Viewer - pdf, djvu, xps and cbz file viewer for android devices
 *
 * Copyright (C) 2011-2013  Michael Bogdanov & Co
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package universe.constellation.orion.viewer

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.view.forEach
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.ViewPager
import com.google.android.material.tabs.TabLayout
import java.util.Date

class OrionHelpActivity : OrionBaseActivity() {

    class InfoFragment : Fragment(R.layout.general_help)

    class AboutFragment : Fragment(R.layout.app_about_fragment)

    class ContributionFragment : Fragment(R.layout.app_contribution_fragment) {
        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)
            val conent = view.findViewById<ViewGroup>(R.id.content)
            conent.forEach { view ->
                if (view is TextView) {
                    view.movementMethod = LinkMovementMethod.getInstance()
                }
            }

            if (Date().before(Date(2024 - 1900, 8, 10))) {
                val survey = view.findViewById<TextView>(R.id.survey)
                val key = resources.getString(R.string.survey_key)
                val fullPath = "https://docs.google.com/forms/d/e/$key/viewform?usp=sf_link"

                val spannable = SpannableStringBuilder(survey.text)
                val onClick = object : ClickableSpan() {
                    override fun onClick(widget: View) {
                        val uri = Uri.parse(fullPath)
                        val intent = Intent(Intent.ACTION_VIEW, uri)
                        startActivity(intent)
                        startActivity(intent)
                    }
                }
                spannable.setSpan(onClick, 0, spannable.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                survey.text = spannable
                survey.movementMethod = LinkMovementMethod.getInstance();
                survey.visibility = View.VISIBLE
            }
        }
    }

    @SuppressLint("MissingSuperCall")
    override fun onCreate(savedInstanceState: Bundle?) {
        onOrionCreate(savedInstanceState, R.layout.app_help_activity, displayHomeAsUpEnabled = true)
        initHelpScreen()
        chooseTab(intent)
    }

    private fun initHelpScreen() {
        val pagerAdapter = HelpSimplePagerAdapter(supportFragmentManager)
        val viewPager = findViewById<ViewPager>(R.id.viewpager)
        viewPager.adapter = pagerAdapter
        val tabLayout = findViewById<View>(R.id.sliding_tabs) as TabLayout
        tabLayout.setupWithViewPager(viewPager)

        val help = tabLayout.getTabAt(0)
        help?.setIcon(R.drawable.new_help)
        help?.setContentDescription(R.string.menu_help_text)

        val about = tabLayout.getTabAt(1)
        about?.setIcon(R.drawable.new_info)
        about?.setContentDescription(R.string.menu_about_text)

        tabLayout.getTabAt(2)?.apply {
            setIcon(R.drawable.contribution)
            setContentDescription(R.string.menu_about_text)
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        chooseTab(intent)
    }

    private fun chooseTab(intent: Intent?) {
        val index = if (intent?.getBooleanExtra(OPEN_ABOUT_TAB, false) == true) 1 else 0
        findViewById<ViewPager>(R.id.viewpager).setCurrentItem(index, false)
    }

    companion object {
        const val OPEN_ABOUT_TAB = "OPEN_ABOUT"
    }

}

internal class HelpSimplePagerAdapter(fm: androidx.fragment.app.FragmentManager) : FragmentStatePagerAdapter(fm) {

    private val fragments: MutableList<Fragment> = arrayListOf(OrionHelpActivity.InfoFragment(), OrionHelpActivity.AboutFragment(), OrionHelpActivity.ContributionFragment())

    override fun getItem(i: Int): Fragment {
        return fragments[i]
    }

    override fun getCount(): Int {
        return fragments.size
    }
}
